heron/api/src/java/org/apache/heron/streamlet/impl/WindowConfigImpl.java

```diff
@@ -19,59 +19,12 @@
 
 package org.apache.heron.streamlet.impl;
 
-import java.time.Duration;
-
 import org.apache.heron.api.bolt.BaseWindowedBolt;
-import org.apache.heron.api.tuple.Tuple;
-import org.apache.heron.api.windowing.EvictionPolicy;
-import org.apache.heron.api.windowing.TriggerPolicy;
 import org.apache.heron.streamlet.WindowConfig;
 
 /**
  * WindowConfigImpl implements the WindowConfig interface.
  */
-public final class WindowConfigImpl implements WindowConfig {
-  private enum WindowType { TIME, COUNT, CUSTOM }
-  private WindowType windowType;
-  private int windowSize;
-  private int slideInterval;
-  private Duration windowDuration;
-  private Duration slidingIntervalDuration;
-  private TriggerPolicy<Tuple, ?> triggerPolicy;
-  private EvictionPolicy<Tuple, ?> evictionPolicy;
-
-  public WindowConfigImpl(Duration windowDuration, Duration slidingIntervalDuration) {
-    this.windowType = WindowType.TIME;
-    this.windowDuration = windowDuration;
-    this.slidingIntervalDuration = slidingIntervalDuration;
-  }
-  public WindowConfigImpl(int windowSize, int slideInterval) {
-    this.windowType = WindowType.COUNT;
-    this.windowSize = windowSize;
-    this.slideInterval = slideInterval;
-  }
-  public WindowConfigImpl(TriggerPolicy<Tuple, ?> triggerPolicy,
-                          EvictionPolicy<Tuple, ?> evictionPolicy) {
-    this.windowType = WindowType.CUSTOM;
-    this.triggerPolicy = triggerPolicy;
-    this.evictionPolicy = evictionPolicy;
-  }
-
-  public void attachWindowConfig(BaseWindowedBolt bolt) {
-    switch(windowType) {
-      case COUNT:
-        bolt.withWindow(BaseWindowedBolt.Count.of(windowSize),
-                        BaseWindowedBolt.Count.of(slideInterval));
-        break;
-      case TIME:
-        bolt.withWindow(windowDuration, slidingIntervalDuration);
-        break;
-      case CUSTOM:
-        bolt.withCustomEvictor(evictionPolicy);
-        bolt.withCustomTrigger(triggerPolicy);
-        break;
-      default:
-        throw new RuntimeException("Unknown windowType " + String.valueOf(windowType));
-    }
-  }
+public abstract class WindowConfigImpl implements WindowConfig {
```

Comment:
Since it's no longer an Impl class, could we move the abstract method to the interface itself and get rid of this class?

Reply:
I tried and it doesn't work. Let me try again.

Reply:
examples/src/java/org/apache/heron/examples/streamlet/ImpressionsAndClicksTopology.java:177: error: reference to reduceByKeyAndWindow is ambiguous
        .reduceByKeyAndWindow(
        ^
  both method <K#1,V>reduceByKeyAndWindow(SerializableFunction<R,K#1>,SerializableFunction<R,V>,WindowConfig,SerializableBinaryOperator<V>) in Streamlet a
nd method <K#2,T>reduceByKeyAndWindow(SerializableFunction<R,K#2>,WindowConfig,T,SerializableBiFunction<T,R,? extends T>) in Streamlet match
  where K#1,V,R,K#2,T are type-variables:
    K#1 extends Object declared in method <K#1,V>reduceByKeyAndWindow(SerializableFunction<R,K#1>,SerializableFunction<R,V>,WindowConfig,SerializableBinar
yOperator<V>)
    V extends Object declared in method <K#1,V>reduceByKeyAndWindow(SerializableFunction<R,K#1>,SerializableFunction<R,V>,WindowConfig,SerializableBinaryOperator<V>)
    R extends Object declared in interface Streamlet
    K#2 extends Object declared in method <K#2,T>reduceByKeyAndWindow(SerializableFunction<R,K#2>,WindowConfig,T,SerializableBiFunction<T,R,? extends T>)
    T extends Object declared in method <K#2,T>reduceByKeyAndWindow(SerializableFunction<R,K#2>,WindowConfig,T,SerializableBiFunction<T,R,? extends T>)

Reply:
Hmm, looks like the compiler can't tell the difference between an interface and lambda expressions at compile time. We can combine `WindowConfig` and `WindowConfigImpl` still as one abstract class to remove redundancy.

Reply:
+1 for reducing the redundancy.